File: Services/pkg/auth/repository.go
package auth

import (
	"fmt"
	"time"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v4"
	helpers "github.com/zercle/gofiber-helpers"
)

type authRepository struct {
	resources models.Resources
}

func NewAuthRepository(resources models.Resources) domain.AuthRepository {
	return &authRepository{resources: resources}
}
func (r *authRepository) SignToken(user models.User, host string, ttl time.Duration) (string, *helpers.ResponseError) {
	token := jwt.NewWithClaims(r.resources.JwtResources.JwtSigningMethod, &jwt.RegisteredClaims{})
	claims := token.Claims.(*jwt.RegisteredClaims)
	claims.Subject = fmt.Sprintf("%d", user.ID)
	claims.Issuer = host
	claims.ExpiresAt = jwt.NewNumericDate(time.Now().Add(ttl))
	signToken, err := token.SignedString(r.resources.JwtResources.JwtSignKey)
	if err != nil {
		return "", &helpers.ResponseError{
			Message: err.Error(),
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Failed to sign json web token",
		}
	}
	return signToken, nil
}

func (r *authRepository) SaveRefreshToken(userID uint, token string, ttl time.Duration) *helpers.ResponseError {
	key := fmt.Sprintf("refresh:%d", userID)
	if err := r.resources.RedisStorage.Set(key, []byte(token), ttl); err != nil {
		return &helpers.ResponseError{
			Message: err.Error(),
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Failed to save refresh token",
		}
	}
	return nil
}

func (r *authRepository) GetRefreshToken(userID uint) (string, *helpers.ResponseError) {
	key := fmt.Sprintf("refresh:%d", userID)
	tokenBytes, err := r.resources.RedisStorage.Get(key)
	if err != nil {
		return "", &helpers.ResponseError{
			Message: err.Error(),
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Failed to get refresh token",
		}
	}
	if tokenBytes == nil {
		return "", &helpers.ResponseError{
			Message: "Refresh token not found",
			Code:    fiber.StatusUnauthorized,
			Source:  helpers.WhereAmI(),
			Title:   "Unauthorized",
		}
	}
	return string(tokenBytes), nil
}

func (r *authRepository) DeleteRefreshToken(userID uint) *helpers.ResponseError {
	key := fmt.Sprintf("refresh:%d", userID)
	if err := r.resources.RedisStorage.Delete(key); err != nil {
		return &helpers.ResponseError{
			Message: err.Error(),
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Failed to delete refresh token",
		}
	}
	return nil
}

func (r *authRepository) ParseToken(tokenString string) (*jwt.Token, *helpers.ResponseError) {
	token, err := jwt.ParseWithClaims(tokenString, &jwt.RegisteredClaims{}, r.resources.JwtResources.JwtKeyfunc)
	if err != nil {
		return nil, &helpers.ResponseError{
			Message: err.Error(),
			Code:    fiber.StatusUnauthorized,
			Source:  helpers.WhereAmI(),
			Title:   "Invalid token",
		}
	}
	return token, nil
}

File: hono-clean-architecture/pkg/auth/repository.ts
import Redis from "ioredis";
import { User, UserResponse, Resources, CacheKeys } from "../models";
import { AuthRepository as IAuthRepository } from "../domain";
import { ResponseError, createError, whereAmI } from "../../internal/infrastructure/custom_error";
import { signJwt, parseJwt, JwtResources } from "../../internal/datasources/jwt";
import { JWTPayload } from "jose";

export class AuthRepository implements IAuthRepository {
  private jwtResources: JwtResources;
  private redisStorage: Redis | null;

  constructor(resources: Resources) {
    this.jwtResources = resources.jwtResources;
    this.redisStorage = resources.redisStorage;
  }

  async signToken(
    user: User | UserResponse,
    host: string,
    ttlSeconds: number,
    roles: string[] = [],
    permissions: string[] = []
  ): Promise<{ token: string | null; error: ResponseError | null }> {
    try {
      const now = Math.floor(Date.now() / 1000);
      const payload: JWTPayload = {
        sub: String(user.id),
        iss: host,
        exp: now + ttlSeconds,
        iat: now,
        roles,
        permissions,
      };

      const token = await signJwt(this.jwtResources, payload);
      return { token, error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Failed to sign token";
      return {
        token: null,
        error: createError(500, whereAmI(), "Failed to sign json web token", message),
      };
    }
  }

  async saveRefreshToken(
    userId: number,
    token: string,
    ttlSeconds: number
  ): Promise<ResponseError | null> {
    if (!this.redisStorage) {
      return createError(500, whereAmI(), "Redis Error", "Redis storage not available");
    }

    try {
      const key = `${CacheKeys.RefreshToken}:${userId}`;
      await this.redisStorage.set(key, token, "EX", ttlSeconds);
      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Failed to save refresh token";
      return createError(500, whereAmI(), "Failed to save refresh token", message);
    }
  }

  async getRefreshToken(
    userId: number
  ): Promise<{ token: string | null; error: ResponseError | null }> {
    if (!this.redisStorage) {
      return {
        token: null,
        error: createError(500, whereAmI(), "Redis Error", "Redis storage not available"),
      };
    }

    try {
      const key = `${CacheKeys.RefreshToken}:${userId}`;
      const token = await this.redisStorage.get(key);

      if (!token) {
        return {
          token: null,
          error: createError(401, whereAmI(), "Unauthorized", "Refresh token not found"),
        };
      }

      return { token, error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Failed to get refresh token";
      return {
        token: null,
        error: createError(500, whereAmI(), "Failed to get refresh token", message),
      };
    }
  }

  async deleteRefreshToken(userId: number): Promise<ResponseError | null> {
    if (!this.redisStorage) {
      return createError(500, whereAmI(), "Redis Error", "Redis storage not available");
    }

    try {
      const key = `${CacheKeys.RefreshToken}:${userId}`;
      await this.redisStorage.del(key);
      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Failed to delete refresh token";
      return createError(500, whereAmI(), "Failed to delete refresh token", message);
    }
  }

  async parseToken(tokenString: string): Promise<{
    payload: JWTPayload | null;
    valid: boolean;
    error: ResponseError | null;
  }> {
    try {
      const { payload, valid } = await parseJwt(this.jwtResources, tokenString);
      return { payload, valid, error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Invalid token";
      return {
        payload: null,
        valid: false,
        error: createError(401, whereAmI(), "Invalid token", message),
      };
    }
  }
}

export function newAuthRepository(resources: Resources): IAuthRepository {
  return new AuthRepository(resources);
}
